mslice/models/workspacemanager/mantid_workspace_provider.py

```diff
@@ -36,14 +44,67 @@ def _processEfixed(self, workspace):
         except RuntimeError:
             self._EfDefined[ws_name] = False
 
+    def _processLoadedWSLimits(self, workspace):
+        """ Processes an (angle-deltaE) workspace to get the limits and step size in angle, energy and |Q| """
+        ws_name = workspace if isinstance(workspace, basestring) else self.get_workspace_name(workspace)
+        ws_h = self.get_workspace_handle(workspace)
+        # For cases, e.g. indirect, where EFixed has not been set yet, return calculate later.
+        try:
+            efix = ws_h.getEFixed(ws_h.getDetector(0).getID())
+        except RuntimeError:   # Efixed not defined
+            # This could occur for malformed NXSPE without the instrument name set.
+            # LoadNXSPE then sets EMode to 'Elastic' and getEFixed fails.
+            if ws_h.run().hasProperty('Ei'):
+                efix = ws_h.run().getProperty('Ei').value
+            else:
+                self._limits[ws_name] = None
+                return
+        except AttributeError: # Wrong workspace type (e.g. cut)
+            self._limits[ws_name] = {}
+            return
+        # Checks that loaded data is in energy transfer.
+        enAxis = ws_h.getAxis(0)
+        if 'DeltaE' not in enAxis.getUnit().unitID():
+            self._limits[ws_name] = None
+            return
+        en = ws_h.getAxis(0).extractValues()
+        # Don't parse all spectra in cases where there are alot to save time.
+        numHist = ws_h.getNumberHistograms()
+        if numHist > 500:
+            theta = [ws_h.detectorTwoTheta(ws_h.getDetector(i)) for i in range(0, numHist, int(numHist/500))]
+        else:
+            theta = [ws_h.detectorTwoTheta(ws_h.getDetector(i)) for i in range(numHist)]
+        # Defines some conversion factors
+        E2q = 2. * constants.m_n / (constants.hbar ** 2)  # Energy to (neutron momentum)^2 (==2m_n/hbar^2)
+        meV2J = constants.e / 1000.                       # meV to Joules
+        m2A = 1.e10                                       # metres to Angstrom
+        if ws_name not in self._limits.keys():
+            self._limits[ws_name] = {}
+        th = np.array([np.min(theta), np.max(theta), np.mean(np.diff(theta))])
+        # Use |Q| at elastic line to get minimum and step size
+        qmin, qmax, qstep = tuple(np.sqrt(E2q * 2 * efix * (1 - np.cos(th)) * meV2J) / m2A)
+        # Use minimum energy (Direct geometry) or maximum energy (Indirect) to get qmax
+        emax = -np.min(en) if (str(ws_h.getEMode()) == 'Direct') else np.max(en)
+        qmax = np.sqrt(E2q * (2 * efix + emax - 2 * np.sqrt(efix * (efix + emax)) * np.cos(th[1])) * meV2J) / m2A
+        self._limits[ws_name]['MomentumTransfer'] = [qmin - qstep, qmax + qstep, qstep]
```

Comment:
Actually, they are defined as globals in https://github.com/mantidproject/mslice/blob/master/mslice/models/projection/powder/mantid_projection_calculator.py#L9 - however it's just file. We would need a C type `include`  file - a class of definitions maybe?

One thing which I found is that using `SofQW3` to convert the data from `Theta` to Q changes the Q unit to `MomentumTransfer`. However, using `ConvertToMD` with the `|Q|` option changes the unit to `|Q|`. Hence in the code, I had to make two entries in the dictionary... I think it would need some changes to the Mantid code to rationalise this, if it's considered a priority. At the moment it's just an annoyance really.